### PR TITLE
Support contour_plot to show parameter relationship

### LINKED ISSUE #26: [Feature] Support contour_plot to show the parameter relationship.
# Feature Request

Porting `plot_contour()` function to TypeScript.
https://github.com/optuna/optuna/blob/v2.5.0/optuna/visualization/_contour.py

### PR DESCRIPTION
**Pull request by 2403hwaseer**
## Contributor License Agreement

This repository (``optuna-dashboard``) and [Goptuna](https://github.com/c-bata/goptuna) share common code.
This pull request may therefore be ported to Goptuna.
Make sure that you understand the consequences concerning licenses and check the box below if you accept the term before creating this pull request.

* [x] I agree this patch may be ported to [Goptuna](https://github.com/c-bata/goptuna) by other Goptuna contributors.

## Reference Issues/PRs

Fixes #26 

## Support Contour Plot to Show parameter relationship

This will help users to see the relationship between parameters in the form of contours on the dashboard.

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -2251,6 +2251,11 @@
       "resolved": "https://registry.npmjs.org/plotly.js-dist/-/plotly.js-dist-1.57.1.tgz",
       "integrity": "sha512-tOOFxNc6PvdDkOUgWCSYc+CcxQULTpaGTROxrtmmF8NuZghcT3EInsToEzhPH3PcGM6bu61fKo73ZiFqxwYtlg=="
     },
+    "plotly.js-dist": {
+      "version": "1.58.4",
+      "resolved": "https://registry.npmjs.org/plotly.js-dist/-/plotly.js-dist-1.58.4.tgz",
+      "integrity": "sha512-oXCTRJFN8FBsHZSQPYoM3LuJQchPUrf6sOXFC0EFdvcr5lmJmLcAsW74jDy9PkRpm3PB+A+2oY1hsUMmk2eZbw=="
+    },
```
Comment (c-bata) [inline package-lock.json:2258]: Why is this change included without the changes of package.json?

File: optuna_dashboard/static/components/GraphContour.tsx [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+    return
+  }
+
+  const layout: Partial<plotly.Layout> = {
+    title: "Contour",
+    margin: {
+      l: 50,
+      r: 50,
+    },
+  }
```
Comment (c-bata) [inline optuna_dashboard/static/components/GraphContour.tsx:127]: Could you explain why this is different with `updateLayout`?

File: optuna_dashboard/static/components/GraphContour.tsx [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+        zerolinecolor: "#f2f5fa",
+        zerolinewidth: 1.5,
+      },
+      yaxis: {
+        gridcolor: "#f2f5fa",
+        gridwidth: 1,
+        zerolinecolor: "#f2f5fa",
+        zerolinewidth: 1.5,
+      },
+      plot_bgcolor: "#E5ecf6",
```
Comment (c-bata) [inline optuna_dashboard/static/components/GraphContour.tsx:164]: Could you explain why this `layout` is different with above?

File: optuna_dashboard/static/components/GraphContour.tsx [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+
+  let paramNames = new Set<string>(trials[0].params.map((p) => p.name))
+  filteredTrials.forEach((t) => {
+    paramNames = new Set<string>(
+      t.params.filter((p) => paramNames.has(p.name)).map((p) => p.name)
+    )
+  })
+
+  if (paramNames.size === 0 || paramNames.size === 1) {
+    plotly.react(plotDomId, [])
```
Comment (c-bata) [inline optuna_dashboard/static/components/GraphContour.tsx:147]: It looks that layout is missing.

File: optuna_dashboard/static/components/GraphContour.tsx [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+    })
+    const paramCategorical = { ...paramValues }
+    const xIndex = paramIndices.indexOf(xAxis)
+    const yIndex = paramIndices.indexOf(yAxis)
+
+    const x_indice = paramCategorical[xIndex].sort((a, b) => (a > b ? 1 : -1))
+    const y_indice = paramCategorical[yIndex].sort((a, b) => (a > b ? 1 : -1))
+
+    const x_indices: string[] = []
+    const y_indices: string[] = []
```
Comment (c-bata) [inline optuna_dashboard/static/components/GraphContour.tsx:200]: Could you avoid using snake_case_with_underscores? Other variables are named in camelCase styles.

**Review by c-bata — CHANGES_REQUESTED** [review]
I left some review comments.

File: package-lock.json (hunk 1 above)
Comment (2403hwaseer) [reply package-lock.json:2258]: Previously, `plotly.js-basic-dist` was being used. However, it did not support contour plots, so I installed `plotly.js-dist` without realising that this particular change had been made in https://github.com/optuna/optuna-dashboard/commit/f8e35c19f715892e5650eb79407801c2ee10bfc4. While resolving the merge conflicts, I forgot to update this. 
I'll make the required change.

File: optuna_dashboard/static/components/GraphContour.tsx (hunk 2 above)
Comment (2403hwaseer) [reply optuna_dashboard/static/components/GraphContour.tsx:127]: I too don't find much use of using two different variables `layout` and `updateLayout`. I'll avoid unnecessary code repitition from next time. Also, I'll make the necessary change.

File: optuna_dashboard/static/components/GraphContour.tsx (hunk 3 above)
Comment (2403hwaseer) [reply optuna_dashboard/static/components/GraphContour.tsx:164]: Actually, I was just checking if this layout was fine and forgot to assign it to a variable. I'll make the change.

File: optuna_dashboard/static/components/GraphContour.tsx (hunk 4 above)
Comment (2403hwaseer) [reply optuna_dashboard/static/components/GraphContour.tsx:147]: Oh yes, I'll make the change.

File: optuna_dashboard/static/components/GraphContour.tsx (hunk 5 above)
Comment (2403hwaseer) [reply optuna_dashboard/static/components/GraphContour.tsx:200]: Yes, sure! I'll adhere to that convention.

File: optuna_dashboard/static/components/GraphContour.tsx [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+  )
+
+  let paramNames = new Set<string>(trials[0].params.map((p) => p.name))
+  filteredTrials.forEach((t) => {
+    paramNames = new Set<string>(
+      t.params.filter((p) => paramNames.has(p.name)).map((p) => p.name)
+    )
+  })
+
+  const paramnames = Array.from(paramNames)
```
Comment (c-bata) [inline optuna_dashboard/static/components/GraphContour.tsx:39]: These variable names (`paramNames` and `paramnames` ) looks confusing to me. Could you refactor this?

File: optuna_dashboard/static/components/GraphContour.tsx [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+  const filteredTrials = trials.filter(
+    (t) => t.state === "Complete" || t.state === "Pruned"
+  )
+
+  let paramNames = new Set<string>(trials[0].params.map((p) => p.name))
+  filteredTrials.forEach((t) => {
+    paramNames = new Set<string>(
+      t.params.filter((p) => paramNames.has(p.name)).map((p) => p.name)
+    )
+  })
```
Comment (c-bata) [inline optuna_dashboard/static/components/GraphContour.tsx:157]: These lines are duplicated to L27-L37. I think we can reduce these computation costs. Could you refactor this?

File: optuna_dashboard/static/components/GraphContour.tsx [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+      })
+      paramValues[i] = valueStrings
+      i++
+    })
+
+    let k = 0
+    paramNames.forEach((paramName) => {
+      paramIndices[k] = paramName
+      k++
+    })
```
Comment (c-bata) [inline optuna_dashboard/static/components/GraphContour.tsx:190]: You can get the index number from the second argument of the callback function.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach

**Review by c-bata — CHANGES_REQUESTED** [review]

File: optuna_dashboard/static/components/GraphContour.tsx (hunk 6 above)
Comment (2403hwaseer) [reply optuna_dashboard/static/components/GraphContour.tsx:39]: Yeah sure!

File: optuna_dashboard/static/components/GraphContour.tsx (hunk 7 above)
Comment (2403hwaseer) [reply optuna_dashboard/static/components/GraphContour.tsx:157]: I tried to remove this part (since it was duplicate) but it was giving an error. However, I'll try to refactor this.

File: optuna_dashboard/static/components/GraphContour.tsx (hunk 8 above)
Comment (2403hwaseer) [reply optuna_dashboard/static/components/GraphContour.tsx:190]: Sounds good! I'll make the required change.

File: optuna_dashboard/static/components/GraphContour.tsx [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+
+  if (trials.length === 0) {
+    plotly.react(plotDomId, [], layout)
+  }
+
+  const filteredTrials = trials.filter(
+    (t) => t.state === "Complete" || t.state === "Pruned"
+  )
+
+  const paramNames = getParamNames(filteredTrials)
```
Comment (c-bata) [inline optuna_dashboard/static/components/GraphContour.tsx:152]: Could you avoid to call getParamNames twice in this component? It requires O(nm) where n is the number of trials and m is the number of parameters.

File: optuna_dashboard/static/components/GraphContour.tsx [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+      zerolinewidth: 1.5,
+    },
+    plot_bgcolor: "#E5ecf6",
+  }
+
+  const trials: Trial[] = study !== null ? study.trials : []
+
+  if (trials.length === 0) {
+    plotly.react(plotDomId, [], layout)
+  }
```
Comment (c-bata) [inline optuna_dashboard/static/components/GraphContour.tsx:146]: I think it's better to check the length of filteredTrials instead of trials.

**Review by c-bata — CHANGES_REQUESTED** [review]

File: optuna_dashboard/static/components/GraphContour.tsx (hunk 9 above)
Comment (2403hwaseer) [reply optuna_dashboard/static/components/GraphContour.tsx]: Yeah sure!

File: optuna_dashboard/static/components/GraphContour.tsx (hunk 10 above)
Comment (2403hwaseer) [reply optuna_dashboard/static/components/GraphContour.tsx]: Yes, sounds good!

**Comment by 2403hwaseer** [comment]
Hi @c-bata , I made the required changes. 
I removed calling `getParamNames` twice. However I am sceptical about my approach. Could you suggest me some other method or is this fine?

**Review by c-bata — CHANGES_REQUESTED** [review]
> I removed calling getParamNames twice. However I am sceptical about my approach. Could you suggest me some other method or is this fine?

Looks good! Thank you for the update. I left an additional review comment.

File: optuna_dashboard/static/components/GraphContour.tsx [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+  const filteredTrials = trials.filter(
+    (t) => t.state === "Complete" || t.state === "Pruned"
+  )
+
+  const paramNames = getParamNames(filteredTrials)
+
+  const classes = useStyles()
+  const [objectiveId, setObjectiveId] = useState<number>(0)
+  const [xAxis, setXAxis] = useState<string>(paramNames[0])
+  const [yAxis, setYAxis] = useState<string>(paramNames[1])
```
[the remark this answers is not in the document]
Comment (2403hwaseer) [reply optuna_dashboard/static/components/GraphContour.tsx]: Yeah sure, I'll fix this.

**Comment by 2403hwaseer** [comment]
Hey @c-bata , can you please review and suggest me if any changes need to be done?

**Comment by c-bata** [comment]
Hi @2403hwaseer

I will merge your changes at #56 with some followup changes 👍
